### PR TITLE
Restrict the serialization

### PR DESCRIPTION
**Pull request by erikzhang**

### CODE REVIEW

File: src/neo/SmartContract/Native/Tokens/Nep5AccountState.cs [hunk 1]
```diff
@@ -29,7 +29,7 @@ protected virtual void FromStruct(Struct @struct)
 
         public byte[] ToByteArray()
         {
-            return StackItemSerializer.Serialize(ToStruct());
+            return StackItemSerializer.Serialize(ToStruct(), 4096);
```
[the remark this answers is not in the document]
Comment (erikzhang) [reply src/neo/SmartContract/Native/Tokens/Nep5AccountState.cs]: There is no difference between `4096` and `int.MaxValue` here.

File: src/neo/SmartContract/InteropService.NEO.cs [hunk 2]
```diff
@@ -387,9 +387,9 @@ private static bool Json_Deserialize(ApplicationEngine engine)
         private static bool Json_Serialize(ApplicationEngine engine)
         {
             var item = engine.CurrentContext.EvaluationStack.Pop();
-            var json = JsonSerializer.Serialize(item);
-
-            engine.CurrentContext.EvaluationStack.Push(json.ToString());
+            var json = JsonSerializer.Serialize(item).ToByteArray(false);
+            if (json.Length > engine.MaxItemSize) return false;
```
[the remark this answers is not in the document]
Comment (erikzhang) [reply src/neo/SmartContract/InteropService.NEO.cs]: I didn't find a easy way to do that.

[the remark this answers is not in the document]
Comment (lock9) [reply src/neo/SmartContract/InteropService.NEO.cs]: Maybe if you just move this check to be inside Serialize is enough. You can pass engine.MaxItemSize as parameter. 
It won't have more performance or anything, but at least we are sure that Serialize is "always safe".

[the remark this answers is not in the document]
Comment (erikzhang) [reply src/neo/SmartContract/InteropService.NEO.cs]: Maybe you can try it. I can't find an easy way.

File: src/neo/SmartContract/StackItemSerializer.cs [hunk 3]
```diff
@@ -177,6 +177,8 @@ private static void Serialize(StackItem item, BinaryWriter writer)
                         writer.Write((byte)StackItemType.Null);
                         break;
                 }
+                if (writer.BaseStream.Position > maxSize)
```
Comment (shargon) [inline src/neo/SmartContract/StackItemSerializer.cs]: I prefer to increment a counter, maybe this BaseStream is not `seekable`

Comment (erikzhang) [reply src/neo/SmartContract/StackItemSerializer.cs]: This is a private method and the `BaseStream` is a `MemoryStream`.

File: src/neo/SmartContract/InteropService.NEO.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (lock9) [reply src/neo/SmartContract/InteropService.NEO.cs]: Thanks, I'm doing that.

[the remark this answers is not in the document]
Comment (lock9) [reply src/neo/SmartContract/InteropService.NEO.cs]: @erikzhang 
Please check https://github.com/neo-project/neo/pull/1355

**Comment by erikzhang** [comment]
@shargon @lock9 Please review it again.

**Comment by shargon** [comment]
Wait, I will add some UT for increase the coverage

**Comment by erikzhang** [comment]
It is a number, if you convert it to a string, you still can't use it in the contract.

**Pull request merged**